File: Snooper/src/main/java/com/prateekj/snooper/networksnooper/fragment/HttpCallFragment.java
package com.prateekj.snooper.networksnooper.fragment;

import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.core.view.MenuItemCompat;
import androidx.core.widget.NestedScrollView;
import androidx.core.widget.NestedScrollView.OnScrollChangeListener;
import androidx.appcompat.widget.SearchView;
import androidx.appcompat.widget.SearchView.OnQueryTextListener;
import android.text.Spannable;
import android.text.SpannableStringBuilder;
import android.text.Spanned;
import android.text.style.BackgroundColorSpan;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.prateekj.snooper.R;
import com.prateekj.snooper.formatter.ResponseFormatterFactory;
import com.prateekj.snooper.infra.BackgroundTaskExecutor;
import com.prateekj.snooper.networksnooper.database.SnooperRepo;
import com.prateekj.snooper.networksnooper.model.Bound;
import com.prateekj.snooper.networksnooper.presenter.HttpCallFragmentPresenter;
import com.prateekj.snooper.networksnooper.viewmodel.HttpBodyViewModel;
import com.prateekj.snooper.networksnooper.views.HttpCallBodyView;
import com.prateekj.snooper.utils.Logger;

import java.util.List;

import static android.view.View.GONE;
import static android.widget.TextView.BufferType.SPANNABLE;
import static com.prateekj.snooper.networksnooper.activity.HttpCallActivity.HTTP_CALL_ID;
import static com.prateekj.snooper.networksnooper.activity.HttpCallActivity.HTTP_CALL_MODE;
import static com.prateekj.snooper.utils.CollectionUtilities.isLast;
import static java.lang.Math.min;

public class HttpCallFragment extends Fragment implements HttpCallBodyView, OnQueryTextListener, OnScrollChangeListener {

  public static final String TAG = HttpCallFragment.class.getSimpleName();
  public static final int NEXT_SET_HIGHLIGHT_SCROLL_LINE_BUFFER = 20;
  public static final int BOUNDS_HIGHLIGHT_SET_SIZE = 50;
  private int mode;
  private HttpBodyViewModel viewModel;
  private HttpCallFragmentPresenter presenter;
  private TextView payloadTextView;
  private NestedScrollView scrollView;
  private int lastBoundHighlightedIndex = 0;
  private List<Bound> bounds;
  private int ythPositionOfLastHighlightedBound;

  @Nullable
  @Override
  public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
    super.onCreateView(inflater, container, savedInstanceState);
    View view = inflater.inflate(R.layout.fragment_response_body, container, false);
    viewModel = new HttpBodyViewModel();
    SnooperRepo repo = new SnooperRepo(getActivity());
    long httpCallId = getArguments().getLong(HTTP_CALL_ID);
    mode = getArguments().getInt(HTTP_CALL_MODE);
    BackgroundTaskExecutor taskExecutor = new BackgroundTaskExecutor(this.getActivity());
    presenter = new HttpCallFragmentPresenter(repo, httpCallId, this, new ResponseFormatterFactory(), taskExecutor);
    setHasOptionsMenu(true);
    return view;
  }

  @Override
  public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
    scrollView = (NestedScrollView) view.findViewById(R.id.scrollView);
    payloadTextView = (TextView) getView().findViewById(R.id.payload_text);
    super.onViewCreated(view, savedInstanceState);
    changeLoaderVisibility(View.VISIBLE);
    presenter.init(viewModel, mode);
    scrollView.setOnScrollChangeListener(this);
  }

  @Override
  public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
    MenuItem searchMenu = menu.findItem(R.id.search_menu);
    searchMenu.setVisible(true);
    ((SearchView) MenuItemCompat.getActionView(searchMenu)).setOnQueryTextListener(this);
    super.onCreateOptionsMenu(menu, inflater);
  }

  @Override
  public void onFormattingDone() {
    SpannableStringBuilder spannableStringBuilder = new SpannableStringBuilder(viewModel.getFormattedBody());
    payloadTextView.setText(spannableStringBuilder, SPANNABLE);
    changeLoaderVisibility(GONE);
  }

  private void changeLoaderVisibility(int visible) {
    getView().findViewById(R.id.embedded_loader).setVisibility(visible);
  }

  @Override
  public void highlightBounds(List<Bound> bounds) {
    this.bounds = bounds;
    Logger.d(TAG, "Total size: " + bounds.size());
    highlightStringFromBounds(bounds.subList(lastBoundHighlightedIndex, min(BOUNDS_HIGHLIGHT_SET_SIZE, bounds.size())));
    scrollTillYOffset(getYthPositionOfBoundInBody(bounds.get(0)));
  }

  @Override
  public void removeOldHighlightedSpans() {
    Spannable spannableString = (Spannable) payloadTextView.getText();
    BackgroundColorSpan[] backgroundSpans = spannableString.getSpans(0, spannableString.length(), BackgroundColorSpan.class);
    for (BackgroundColorSpan span : backgroundSpans) {
      spannableString.removeSpan(span);
    }
  }

  private void scrollTillYOffset(final int yOffset) {
    scrollView.post(new Runnable() {
      @Override
      public void run() {
        scrollView.scrollTo(0, yOffset);
      }
    });
  }

  @Override
  public boolean onQueryTextSubmit(String query) {
    return false;
  }

  @Override
  public boolean onQueryTextChange(String newText) {
    lastBoundHighlightedIndex = 0;
    presenter.searchInBody(newText.toLowerCase());
    return true;
  }

  @Override
  public void onScrollChange(NestedScrollView v, int scrollX, int scrollY, int oldScrollX, int oldScrollY) {
    if (hasBoundsToHighlight() && needToHighlightNextSetOfBounds(scrollY)) {
      final int calculatedToIndex = lastBoundHighlightedIndex + BOUNDS_HIGHLIGHT_SET_SIZE;
      highlightStringFromBounds(bounds.subList(lastBoundHighlightedIndex + 1, min(calculatedToIndex, bounds.size())));
    }
  }

  private int getYthPositionOfBoundInBody(Bound bound) {
    int lineNumber = getLineNumber(bound.getLeft());
    return payloadTextView.getLayout().getLineTop(lineNumber);
  }

  public int getLineNumber(int indexOfFirstOccurrenceWord) {
    return payloadTextView.getLayout().getLineForOffset(indexOfFirstOccurrenceWord);
  }

  private void highlightStringFromBounds(List<Bound> bounds) {
    final Spannable text = (Spannable) payloadTextView.getText();
    payloadTextView.postDelayed(getHighlightAction(text, bounds), 5);
  }

  @NonNull
  private Runnable getHighlightAction(final Spannable text, final List<Bound> boundsCurrentSet) {
    return new Runnable() {
      @Override
      public void run() {
        for (Bound bound : boundsCurrentSet) {
          text.setSpan(new BackgroundColorSpan(getResources().getColor(R.color.snooper_text_highlight_color)), bound.getLeft(), bound.getRight(), Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
          if (isLast(boundsCurrentSet, bound)) {
            ythPositionOfLastHighlightedBound = getYthPositionOfBoundInBody(bound);
            lastBoundHighlightedIndex = bounds.indexOf(bound);
          }
        }
      }
    };
  }


  private boolean needToHighlightNextSetOfBounds(int scrollY) {
    return ythPositionOfLastHighlightedBound - scrollY < NEXT_SET_HIGHLIGHT_SCROLL_LINE_BUFFER;
  }

  private boolean hasBoundsToHighlight() {
    return bounds != null && lastBoundHighlightedIndex < (bounds.size() - 1);
  }
}

File: Snooper/src/main/java/com/prateekj/snooper/networksnooper/fragment/HttpCallFragment.kt
package com.prateekj.snooper.networksnooper.fragment

import android.os.Bundle
import android.text.Spannable
import android.text.SpannableStringBuilder
import android.text.Spanned
import android.text.style.BackgroundColorSpan
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.View
import android.view.View.GONE
import android.view.ViewGroup
import android.widget.TextView.BufferType.SPANNABLE
import androidx.appcompat.widget.SearchView
import androidx.appcompat.widget.SearchView.OnQueryTextListener
import androidx.core.view.MenuItemCompat
import androidx.core.widget.NestedScrollView
import androidx.core.widget.NestedScrollView.OnScrollChangeListener
import androidx.fragment.app.Fragment
import com.prateekj.snooper.R
import com.prateekj.snooper.formatter.ResponseFormatterFactory
import com.prateekj.snooper.infra.BackgroundTaskExecutor
import com.prateekj.snooper.networksnooper.activity.HttpCallActivity.Companion.HTTP_CALL_ID
import com.prateekj.snooper.networksnooper.activity.HttpCallActivity.Companion.HTTP_CALL_MODE
import com.prateekj.snooper.networksnooper.database.SnooperRepo
import com.prateekj.snooper.networksnooper.model.Bound
import com.prateekj.snooper.networksnooper.presenter.HttpCallFragmentPresenter
import com.prateekj.snooper.networksnooper.viewmodel.HttpBodyViewModel
import com.prateekj.snooper.networksnooper.views.HttpCallBodyView
import com.prateekj.snooper.utils.Logger
import kotlinx.android.synthetic.main.fragment_response_body.*
import kotlin.math.min

class HttpCallFragment : Fragment(), HttpCallBodyView, OnQueryTextListener, OnScrollChangeListener {
  private var mode: Int = 0
  private lateinit var viewModel: HttpBodyViewModel
  private var presenter: HttpCallFragmentPresenter? = null
  private var lastBoundHighlightedIndex = 0
  private var bounds: List<Bound>? = null
  private var ythPositionOfLastHighlightedBound: Int = 0

  override fun onCreateView(
    inflater: LayoutInflater,
    container: ViewGroup?,
    savedInstanceState: Bundle?
  ): View? {
    super.onCreateView(inflater, container, savedInstanceState)
    val view = inflater.inflate(R.layout.fragment_response_body, container, false)
    viewModel = HttpBodyViewModel()
    val repo = SnooperRepo(activity!!)
    val httpCallId = arguments!!.getLong(HTTP_CALL_ID)
    mode = arguments!!.getInt(HTTP_CALL_MODE)
    val taskExecutor = BackgroundTaskExecutor(this.activity!!)
    presenter = HttpCallFragmentPresenter(
      repo,
      httpCallId,
      this,
      ResponseFormatterFactory(),
      taskExecutor
    )
    setHasOptionsMenu(true)
    return view
  }

  override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
    super.onViewCreated(view, savedInstanceState)
    changeLoaderVisibility(View.VISIBLE)
    presenter!!.init(viewModel, mode)
    scrollView!!.setOnScrollChangeListener(this)
  }

  override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
    val searchMenu = menu.findItem(R.id.search_menu)
    searchMenu.isVisible = true
    (MenuItemCompat.getActionView(searchMenu) as SearchView).setOnQueryTextListener(this)
    super.onCreateOptionsMenu(menu, inflater)
  }

  override fun onFormattingDone() {
    val spannableStringBuilder = SpannableStringBuilder(viewModel.formattedBody)
    payload_text!!.setText(spannableStringBuilder, SPANNABLE)
    changeLoaderVisibility(GONE)
  }

  private fun changeLoaderVisibility(visible: Int) {
    embedded_loader.visibility = visible
  }

  override fun highlightBounds(bounds: List<Bound>) {
    this.bounds = bounds
    Logger.d(HttpCallFragment::class.java.simpleName, "Total size: " + bounds.size)
    highlightStringFromBounds(
      bounds.subList(
        lastBoundHighlightedIndex,
        min(BOUNDS_HIGHLIGHT_SET_SIZE, bounds.size)
      )
    )
    scrollTillYOffset(getYthPositionOfBoundInBody(bounds[0]))
  }

  override fun removeOldHighlightedSpans() {
    val spannableString = payload_text!!.text as Spannable
    val backgroundSpans =
      spannableString.getSpans(0, spannableString.length, BackgroundColorSpan::class.java)
    for (span in backgroundSpans) {
      spannableString.removeSpan(span)
    }
  }

  private fun scrollTillYOffset(yOffset: Int) {
    scrollView!!.post { scrollView!!.scrollTo(0, yOffset) }
  }

  override fun onQueryTextSubmit(query: String): Boolean {
    return false
  }

  override fun onQueryTextChange(newText: String): Boolean {
    lastBoundHighlightedIndex = 0
    presenter!!.searchInBody(newText.toLowerCase())
    return true
  }

  override fun onScrollChange(
    v: NestedScrollView,
    scrollX: Int,
    scrollY: Int,
    oldScrollX: Int,
    oldScrollY: Int
  ) {
    if (hasBoundsToHighlight() && needToHighlightNextSetOfBounds(scrollY)) {
      val calculatedToIndex = lastBoundHighlightedIndex + BOUNDS_HIGHLIGHT_SET_SIZE
      highlightStringFromBounds(
        bounds!!.subList(
          lastBoundHighlightedIndex + 1,
          min(calculatedToIndex, bounds!!.size)
        )
      )
    }
  }

  private fun getYthPositionOfBoundInBody(bound: Bound): Int {
    val lineNumber = getLineNumber(bound.left)
    return payload_text!!.layout.getLineTop(lineNumber)
  }

  private fun getLineNumber(indexOfFirstOccurrenceWord: Int): Int {
    return payload_text!!.layout.getLineForOffset(indexOfFirstOccurrenceWord)
  }

  private fun highlightStringFromBounds(bounds: List<Bound>) {
    val text = payload_text!!.text as Spannable
    payload_text!!.postDelayed(getHighlightAction(text, bounds), 5)
  }

  private fun getHighlightAction(text: Spannable, boundsCurrentSet: List<Bound>): Runnable {
    return Runnable {
      for (bound in boundsCurrentSet) {
        text.setSpan(
          BackgroundColorSpan(resources.getColor(R.color.snooper_text_highlight_color)),
          bound.left,
          bound.right,
          Spanned.SPAN_EXCLUSIVE_EXCLUSIVE
        )
        if (boundsCurrentSet.last() == bound) {
          ythPositionOfLastHighlightedBound = getYthPositionOfBoundInBody(bound)
          lastBoundHighlightedIndex = bounds!!.indexOf(bound)
        }
      }
    }
  }


  private fun needToHighlightNextSetOfBounds(scrollY: Int): Boolean {
    return ythPositionOfLastHighlightedBound - scrollY < NEXT_SET_HIGHLIGHT_SCROLL_LINE_BUFFER
  }

  private fun hasBoundsToHighlight(): Boolean {
    return bounds != null && lastBoundHighlightedIndex < bounds!!.size - 1
  }

  companion object {
    const val NEXT_SET_HIGHLIGHT_SCROLL_LINE_BUFFER = 20
    const val BOUNDS_HIGHLIGHT_SET_SIZE = 50
  }
}
